Expanded C++ language and feature macros

samsinsane opened the pull request.

jproust reviewed (commented):
I don't see a use of UDCPP14 ?

samsinsane commented:
> I don't see a use of UDCPP14 ?

Yeah, I couldn't be bothered adding a second PR just to add the `[[nodiscard]]` wrapper. I initially just wanted that wrapper, which needed C++17, so I added that. I saw there was an extended version for C++20, so I added that too. We then had C++11, C++17 and C++20 detection, not having C++14 seemed weird, which is also the version that udCore builds against.

`udPlatform.h` should have had all of the language detections from the start. It's whole purpose is to provide these helpers. On a few occasions I've wanted a `UDCPP14` or a `UDCPP17` and ended up just bumping _everything_ in udSDK or udStream to C++17 instead because it was easier.

jproust reviewed (approved):
LGTM
